core/src/server/rpc_module.rs

```diff
@@ -655,38 +657,27 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 			self.methods.mut_callbacks().insert(
 				subscribe_method_name,
 				MethodCallback::new_subscription(Arc::new(move |id, params, method_sink, conn| {
-					let (conn_tx, conn_rx) = oneshot::channel::<()>();
+					let sub_id: RpcSubscriptionId = conn.id_provider.next_id().into_owned();
 
-					let sub_id = {
-						let sub_id: RpcSubscriptionId = conn.id_provider.next_id().into_owned();
-						let uniq_sub = SubscriptionKey { conn_id: conn.conn_id, sub_id: sub_id.clone() };
-
-						subscribers.lock().insert(uniq_sub, (method_sink.clone(), conn_rx));
-
-						sub_id
-					};
-
-					method_sink.send_response(id.clone(), &sub_id);
```

Comment:
unfortunately this will a subscription ID whether to call was actually successful or not

and we can't do this after the `callback` because the subscription sink might already have started sending out stuff which is no go.

so indeed we need the users manually to call `accept` or `reject` the subscription whether call was ok after parsing the params. we can't do it because it depends on their callback.

so I introduced `PendingSubscription` that you need to call `reject` or `accept` on which should work.

Reply:
I'm trying to understand; under which circumstances would the callback _not_ run `PendingSubscription::accept()`? Is this so that a provided subscription callback can do arbitrary processing in order to then decide that theere is some issue and we don't want to "begin" the subscription but instead want to return an error (eg parameters aren't what we expected)?

Reply:
(that being the case; I had a skim over and I like the PendingSubscription stuff; it's a nice way to ensure that the thing is either rejected or accepted before any messages are sent :))

Reply:
> I'm trying to understand; under which circumstances would the callback not run PendingSubscription::accept()

It's a way to ensure to `params.parse()` would fail the in `proc macro code` which does that under the hood and additionally do let the user decide when to accept the subscription as you said.

Reply:
> So, if I understand it right, there are basically 2 places an error can happen with a subscription;

Exactly.

> Before it is accepted (ie the params aren't valid, or some other thing that means it's not valid to ask for said subscription (eg server not configured to allow it or something)). Here, we never want to give back a subscription ID etc and want to bail with something like an { jsonrpc: "2.0", error: { code: 123, message: "Invalid params" }} style message, I guess?

Yes, before a subscription call with invalid params was replied to twice because of some of bug in the logic. Because the actual callback (user defined) we can't know whether the params were valid or not unless we could pause the mpsc channel but that seems like a bad idea. So instead we have this `accept/reject API` that user has to use. Additionally I have a drop impl for this `PendingSubscription` so the actual call will be rejected if this is dropped. 

> During the subscription. Some error could occur mid-subscription that, I assume, means we want to end the subscription and report the error back (would we always want to end the subscription on an error?). I guess in this case we want to return a message like { jsonrpsee: "2.0", method: "subscribe_foo", params: { subscription: "123abc", error: { code: 124, message: "Something went wrong" } } }.

> I guess this change allows for (1). Is there a way for people to do (2)?

Yes, there is the `close API` where folks can close down the subscription using the `jsonrpsee::Error`.
Before, jsonrpsee closed down subscription and sent out this under the hood by I have realized that this may not be needed for some use cases such as if subscription method itself has defined some state machine or something for this themselves then we force this overhead on them. Also not break comparability with some libraries "may" not support this.

> We also want a way to signal that the subscription has closed without an error. I guess the options for this are:

> Return an error (like in (2) above) with a code/message that signals the thing has closed.
    Invent more syntax like params: { subscription: "123abc", closed: true } to signal this.
    Leave it up to users to interpret a certain non-error response as meaning that no further messages will be yielded.

Currently, we treat "closed" as an error on send out the same format with some other status code that we pre-defined:

```
Notification { jsonrpc: TwoPointZero, method: "n", params: SubscriptionPayloadError { subscription: Num(60482361782361), error: Object({"code": Number(-32003), "message": String("Subscription was closed by server: Success")}) }
```

So that's is a bit annoying but I think that's reasonable workaround not to break comparability with polkadot JS for now, however this now possible to opt in/out from one needs to call `close` explictly now:  

`let _ = sink.pipe_from_try_stream(stream).await.map_err(|e| sink.close(e));`   

> How did JSONRPC handle this sort of thing? Are we leaning towards doing (1)?

JSON-RPC doesn't support `close notifications` AFAIU but it (1) similar to what is implemented in this PR.
So I think it relies on the impl to have definition when to close the subscriptions using ordinary notifications.

Reply:
Thank you, that's really useful info!

To summarize, with this PR:
- We can reject a pending subscription with a `{ jsonrpc: "2.0", error: {...} }` style error of our choice using `pending.reject(error)`.
- We can end a subscription at any time with `{ jsonrpsee: "2.0", method: "subscribe_foo", params: { subscription: "123abc", error: { ... } } }` style error using `sink.close(error)`.
- Subscriptions that are dropped are now silently closed (for compatibility reasons, incase subscriptions want to signal that they are closed via the "success" messages and not via a "subscription closed" error). In other words, we enforce no opinion on what happens when a subscription is closed, and leave it up to the subscription itself to decide whether to emit an "error" or communicate closure via some other message.

I think if I've understood all of that right, it sounds good to me! The more we stay compatible with how jsonrpc works, the easier!

Reply:
yeah, I think that should correct after the latest changes I pushed for a short while ago.

The only thing is that the subscription API is bit verbose now to work with but it should work.
It might even make sense to do the jsonrpc approach and implement `SinkExt`  on the SubscriptionSink to not have to deal with all this.